### PR TITLE
core: native serialisation

### LINKED ISSUE #1501: Implement native contract serialization changes
See neo-project/neo#1988.

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #1501

### CODE REVIEW

File: pkg/core/native/oracle.go [hunk 1]
```diff
@@ -329,6 +332,9 @@ func (o *Oracle) PutRequestInternal(id uint64, req *state.OracleRequest, d dao.D
 	if err := o.getSerializableFromDAO(d, key, lst); err != nil && !errors.Is(err, storage.ErrKeyNotFound) {
 		return err
 	}
+	if len(*lst) > maxRequestsCount {
+		return fmt.Errorf("there are too many pending responces for %s url", req.URL)
```
Comment (roman-khimov) [inline pkg/core/native/oracle.go:336]: Too many pending requests probably.

File: pkg/core/native/oracle.go [hunk 2]
```diff
@@ -329,6 +332,9 @@ func (o *Oracle) PutRequestInternal(id uint64, req *state.OracleRequest, d dao.D
 	if err := o.getSerializableFromDAO(d, key, lst); err != nil && !errors.Is(err, storage.ErrKeyNotFound) {
 		return err
 	}
+	if len(*lst) > maxRequestsCount {
```
Comment (roman-khimov) [inline pkg/core/native/oracle.go:335]: `>=`

File: pkg/core/native/policy.go [hunk 3]
```diff
@@ -43,8 +41,8 @@ var (
 	// feePerByteKey is a key used to store the minimum fee per byte for
 	// transaction.
 	feePerByteKey = []byte{10}
-	// blockedAccountsKey is a key used to store the list of blocked accounts.
-	blockedAccountsKey = []byte{15}
+	// blockedAccountPrefix is a prefix used to store blocked account.
+	blockedAccountPrefix = []byte{15}
```
Comment (roman-khimov) [inline pkg/core/native/policy.go:45]: Prefixes are constants usually.

File: pkg/core/native/policy.go [hunk 4]
```diff
@@ -64,7 +62,7 @@ type Policy struct {
 	feePerByte              int64
 	maxBlockSystemFee       int64
 	maxVerificationGas      int64
-	blockedAccounts         BlockedAccounts
+	blockedAccounts         map[util.Uint160]bool
```
Comment (roman-khimov) [inline pkg/core/native/policy.go:65]: I don't really like this change, this cache could still easily be `[]util.Uint160` with the new storage model, maps are a bit more heavy for this tasks. Though in practice it very well may be completely irrelevant.

File: pkg/core/native/policy.go [hunk 5]
```diff
@@ -306,32 +301,26 @@ func (p *Policy) GetMaxBlockSystemFeeInternal(dao dao.DAO) int64 {
...
 
-// GetBlockedAccountsInternal returns list of blocked accounts hashes.
-func (p *Policy) GetBlockedAccountsInternal(dao dao.DAO) (BlockedAccounts, error) {
+// IsBlockedInternal checks whether provided account is blocked
+func (p *Policy) IsBlockedInternal(dao dao.DAO, hash util.Uint160) bool {
 	p.lock.RLock()
 	defer p.lock.RUnlock()
 	if p.isValid {
-		return p.blockedAccounts, nil
+		_, ok := p.blockedAccounts[hash]
```
Comment (roman-khimov) [inline pkg/core/native/policy.go:315]: `return p.blockedAccounts[hash]`?

File: pkg/core/native/policy.go (hunk 4 above)
Comment (roman-khimov) [reply pkg/core/native/policy.go:65]: And it's simpler this way, of course.

File: pkg/core/native/native_neo.go [hunk 6]
```diff
@@ -189,10 +191,15 @@ func (n *NEO) Initialize(ic *interop.Context) error {
 	}
 	n.mint(ic, h, big.NewInt(NEOTotalSupply))
 
-	gr := &state.GASRecord{{Index: 0, GASPerBlock: *big.NewInt(5 * GASFactor)}}
+	var index uint32 = 0
+	value := big.NewInt(5 * GASFactor)
+	key := make([]byte, 5)
+	key[0] = prefixGASPerBlock
+	binary.BigEndian.PutUint32(key[1:], index)
+	gr := &state.GASRecord{{Index: index, GASPerBlock: *big.NewInt(5 * GASFactor)}}
```
Comment (roman-khimov) [inline pkg/core/native/native_neo.go:199]: Reuse `value` here?

File: pkg/core/native/native_neo.go [hunk 7]
```diff
@@ -189,10 +191,15 @@ func (n *NEO) Initialize(ic *interop.Context) error {
 	}
 	n.mint(ic, h, big.NewInt(NEOTotalSupply))
 
-	gr := &state.GASRecord{{Index: 0, GASPerBlock: *big.NewInt(5 * GASFactor)}}
+	var index uint32 = 0
+	value := big.NewInt(5 * GASFactor)
+	key := make([]byte, 5)
+	key[0] = prefixGASPerBlock
+	binary.BigEndian.PutUint32(key[1:], index)
```
Comment (roman-khimov) [inline pkg/core/native/native_neo.go:198]: Some helper for the key?

File: pkg/core/state/native_state.go [hunk 8]
```diff
@@ -23,75 +22,15 @@ type NEOBalanceState struct {
 }
 
 // GASIndexPair contains block index together with generated gas per block.
+// It is used only by NEO cache.
```
Comment (roman-khimov) [inline pkg/core/state/native_state.go:25]: Then it should be moved out of the `state` to being internal `native` type.

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/core/native/policy.go (hunk 4 above)
Comment (AnnaShaleva) [reply pkg/core/native/policy.go]: Replaced with array.

File: pkg/core/native/policy.go (hunk 5 above)
Comment (AnnaShaleva) [reply pkg/core/native/policy.go]: It's array now, so replaced with `range`.

File: pkg/core/native/native_neo.go (hunk 7 above)
Comment (AnnaShaleva) [reply pkg/core/native/native_neo.go]: United this part of code with `putBigInteger`, now we have `putGASRecord` method.

File: pkg/core/native/policy.go [hunk 9]
```diff
@@ -220,15 +213,18 @@ func (p *Policy) OnPersistEnd(dao dao.DAO) error {
...
+		return fmt.Errorf("failed to get blocked accounts from storage: %w", err)
+	}
+	for key := range siMap {
+		hash, err := util.Uint160DecodeBytesBE([]byte(key))
+		if err != nil {
+			return fmt.Errorf("failed to decode blocked account hash: %w", err)
+		}
+		p.blockedAccounts = append(p.blockedAccounts, hash)
 	}
-	p.blockedAccounts = ba
```
Comment (roman-khimov) [inline pkg/core/native/policy.go:231]: `sort.Slice(p.blockedAccounts)`

File: pkg/core/native/policy.go [hunk 10]
```diff
@@ -306,32 +302,26 @@ func (p *Policy) GetMaxBlockSystemFeeInternal(dao dao.DAO) int64 {
...
+		for _, ba := range p.blockedAccounts {
+			if hash.Equals(ba) {
+				return true
+			}
+		}
+		return false
 	}
-	return ba, nil
+	key := append([]byte{blockedAccountPrefix}, hash.BytesBE()...)
+	return dao.GetStorageItem(p.ContractID, key) != nil
```
Comment (roman-khimov) [inline pkg/core/native/policy.go:324]: `sort.Search`

Comment (roman-khimov) [reply pkg/core/native/policy.go:324]: Like it was in `CheckPolicy` before this change.

File: pkg/core/native/gas_record.go [hunk 11]
```diff
@@ -0,0 +1,13 @@
+package native
+
+import "math/big"
+
+// GASIndexPair contains block index together with generated gas per block.
+// It is used to cache NEO GASRecords.
+type GASIndexPair struct {
```
Comment (roman-khimov) [inline pkg/core/native/gas_record.go:7]: Does it need to be exported?

File: pkg/core/native/gas_record.go [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+	Index       uint32
+	GASPerBlock big.Int
+}
+
+// GASRecord contains history of gas per block changes. It is used only by NEO cache.
+type GASRecord []GASIndexPair
```
Comment (roman-khimov) [inline pkg/core/native/gas_record.go:13]: Does it need to be exported?

File: pkg/core/native/native_neo.go [hunk 13]
```diff
@@ -437,9 +451,8 @@ func (n *NEO) CalculateBonus(ic *interop.Context, value *big.Int, start, end uin
 	} else if value.Sign() < 0 {
 		return nil, errors.New("negative value")
 	}
-	si := ic.DAO.GetStorageItem(n.ContractID, []byte{prefixGASPerBlock})
-	var gr state.GASRecord
-	if err := gr.FromBytes(si.Value); err != nil {
+	gr, err := n.getSortedGASRecordFromDAO(ic.DAO)
```
Comment (roman-khimov) [inline pkg/core/native/native_neo.go:454]: Why doesn't it use `n.gasPerBlock` here?

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/core/native/gas_record.go (hunk 11 above)
Comment (AnnaShaleva) [reply pkg/core/native/gas_record.go]: No, we don't use it anywhere else. Fixed.

File: pkg/core/native/native_neo.go (hunk 13 above)
Comment (AnnaShaleva) [reply pkg/core/native/native_neo.go]: I checked all users of `CalculateBonus`, so it actually can use cached values. Fixed.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**